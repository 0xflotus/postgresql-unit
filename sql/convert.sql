-- length
WITH
  l(u) AS (VALUES
	('mm'),
	('m'),
	('km'),
	('in'),
	('ft'),
	('yd'),
	('mi'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview

-- area
WITH
  l(u) AS (VALUES
	('mm^2'),
	('m^2'),
	('km^2'),
	('in^2'),
	('ft^2'),
	('yd^2'),
	('mi^2'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview

-- volume
WITH
  l(u) AS (VALUES
	('mm^3'),
	('l'),
	('m^3'),
	('in^3'),
	('ft^3'),
	('yd^3'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview

-- weight
WITH
  l(u) AS (VALUES
	('g'),
	('kg'),
	('t'),
	('oz'),
	('lb'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview

-- time
WITH
  l(u) AS (VALUES
	('s'),
	('min'),
	('h'),
	('d'),
	('julianyear'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview
