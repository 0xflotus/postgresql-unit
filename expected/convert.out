-- length
WITH
  l(u) AS (VALUES
	('mm'),
	('m'),
	('km'),
	('in'),
	('ft'),
	('yd'),
	('mi'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview
 u  |     mm     |     m      |      km      |          in           |           ft           |           yd           |           mi            
----+------------+------------+--------------+-----------------------+------------------------+------------------------+-------------------------
 mm | 1 mm       | 0.001 m    | 1e-06 km     | 0.0393700787401575 in | 0.00328083989501312 ft | 0.00109361329833771 yd | 6.21371192237334e-07 mi
 m  | 1000 mm    | 1 m        | 0.001 km     | 39.3700787401575 in   | 3.28083989501312 ft    | 1.09361329833771 yd    | 0.000621371192237334 mi
 km | 1000000 mm | 1000 m     | 1 km         | 39370.0787401575 in   | 3280.83989501312 ft    | 1093.61329833771 yd    | 0.621371192237334 mi
 in | 25.4 mm    | 0.0254 m   | 2.54e-05 km  | 1 in                  | 0.0833333333333333 ft  | 0.0277777777777778 yd  | 1.57828282828283e-05 mi
 ft | 304.8 mm   | 0.3048 m   | 0.0003048 km | 12 in                 | 1 ft                   | 0.333333333333333 yd   | 0.000189393939393939 mi
 yd | 914.4 mm   | 0.9144 m   | 0.0009144 km | 36 in                 | 3 ft                   | 1 yd                   | 0.000568181818181818 mi
 mi | 1609344 mm | 1609.344 m | 1.609344 km  | 63360 in              | 5280 ft                | 1760 yd                | 1 mi
(7 rows)

WITH
  l(u) AS (VALUES
	('mm'),
	('m'),
	('km'),
	('in'),
	('ft'),
	('yd'),
	('mi'))
SELECT
  l1.u, l2.u, l1.u::unit @@ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview
 u  |   mm    |    m     |    km     |         in         |         ft          |         yd          |          mi          
----+---------+----------+-----------+--------------------+---------------------+---------------------+----------------------
 mm |       1 |    0.001 |     1e-06 | 0.0393700787401575 | 0.00328083989501312 | 0.00109361329833771 | 6.21371192237334e-07
 m  |    1000 |        1 |     0.001 |   39.3700787401575 |    3.28083989501312 |    1.09361329833771 | 0.000621371192237334
 km | 1000000 |     1000 |         1 |   39370.0787401575 |    3280.83989501312 |    1093.61329833771 |    0.621371192237334
 in |    25.4 |   0.0254 |  2.54e-05 |                  1 |  0.0833333333333333 |  0.0277777777777778 | 1.57828282828283e-05
 ft |   304.8 |   0.3048 | 0.0003048 |                 12 |                   1 |   0.333333333333333 | 0.000189393939393939
 yd |   914.4 |   0.9144 | 0.0009144 |                 36 |                   3 |                   1 | 0.000568181818181818
 mi | 1609344 | 1609.344 |  1.609344 |              63360 |                5280 |                1760 |                    1
(7 rows)

-- area
WITH
  l(u) AS (VALUES
	('mm^2'),
	('m^2'),
	('km^2'),
	('in^2'),
	('ft^2'),
	('yd^2'),
	('mi^2'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview
  u   |        mm^2        |        m^2         |        km^2         |          in^2           |           ft^2            |           yd^2            |           mi^2            
------+--------------------+--------------------+---------------------+-------------------------+---------------------------+---------------------------+---------------------------
 mm^2 | 1 mm^2             | 1e-06 m^2          | 1e-12 km^2          | 0.0015500031000062 in^2 | 1.07639104167097e-05 ft^2 | 1.19599004630108e-06 yd^2 | 3.86102158542446e-13 mi^2
 m^2  | 1000000 mm^2       | 1 m^2              | 1e-06 km^2          | 1550.0031000062 in^2    | 10.7639104167097 ft^2     | 1.19599004630108 yd^2     | 3.86102158542446e-07 mi^2
 km^2 | 1000000000000 mm^2 | 1000000 m^2        | 1 km^2              | 1550003100.0062 in^2    | 10763910.4167097 ft^2     | 1195990.04630108 yd^2     | 0.386102158542446 mi^2
 in^2 | 645.16 mm^2        | 0.00064516 m^2     | 6.4516e-10 km^2     | 1 in^2                  | 0.00694444444444444 ft^2  | 0.000771604938271605 yd^2 | 2.49097668605244e-10 mi^2
 ft^2 | 92903.04 mm^2      | 0.09290304 m^2     | 9.290304e-08 km^2   | 144 in^2                | 1 ft^2                    | 0.111111111111111 yd^2    | 3.58700642791552e-08 mi^2
 yd^2 | 836127.36 mm^2     | 0.83612736 m^2     | 8.3612736e-07 km^2  | 1296 in^2               | 9 ft^2                    | 1 yd^2                    | 3.22830578512397e-07 mi^2
 mi^2 | 2589988110336 mm^2 | 2589988.110336 m^2 | 2.589988110336 km^2 | 4014489600 in^2         | 27878400 ft^2             | 3097600 yd^2              | 1 mi^2
(7 rows)

-- volume
WITH
  l(u) AS (VALUES
	('mm^3'),
	('l'),
	('m^3'),
	('in^3'),
	('ft^3'),
	('yd^3'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview
  u   |        mm^3        |        l        |        m^3         |           in^3            |           ft^3            |           yd^3            
------+--------------------+-----------------+--------------------+---------------------------+---------------------------+---------------------------
 mm^3 | 1 mm^3             | 1e-06 l         | 1e-09 m^3          | 6.10237440947323e-05 in^3 | 3.53146667214886e-08 ft^3 | 1.30795061931439e-09 yd^3
 l    | 1000000 mm^3       | 1 l             | 0.001 m^3          | 61.0237440947323 in^3     | 0.0353146667214886 ft^3   | 0.00130795061931439 yd^3
 m^3  | 1000000000 mm^3    | 1000 l          | 1 m^3              | 61023.7440947323 in^3     | 35.3146667214886 ft^3     | 1.30795061931439 yd^3
 in^3 | 16387.064 mm^3     | 0.016387064 l   | 1.6387064e-05 m^3  | 1 in^3                    | 0.000578703703703704 ft^3 | 2.14334705075446e-05 yd^3
 ft^3 | 28316846.592 mm^3  | 28.316846592 l  | 0.028316846592 m^3 | 1728 in^3                 | 1 ft^3                    | 0.037037037037037 yd^3
 yd^3 | 764554857.984 mm^3 | 764.554857984 l | 0.764554857984 m^3 | 46656 in^3                | 27 ft^3                   | 1 yd^3
(6 rows)

-- weight
WITH
  l(u) AS (VALUES
	('g'),
	('kg'),
	('t'),
	('oz'),
	('lb'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview
 u  |       g        |        kg         |         t          |          oz           |           lb           
----+----------------+-------------------+--------------------+-----------------------+------------------------
 g  | 1 g            | 0.001 kg          | 1e-06 t            | 0.0352739619495804 oz | 0.00220462262184878 lb
 kg | 1000 g         | 1 kg              | 0.001 t            | 35.2739619495804 oz   | 2.20462262184878 lb
 t  | 1000000 g      | 1000 kg           | 1 t                | 35273.9619495804 oz   | 2204.62262184878 lb
 oz | 28.349523125 g | 0.028349523125 kg | 2.8349523125e-05 t | 1 oz                  | 0.0625 lb
 lb | 453.59237 g    | 0.45359237 kg     | 0.00045359237 t    | 16 oz                 | 1 lb
(5 rows)

-- time
WITH
  l(u) AS (VALUES
	('s'),
	('min'),
	('h'),
	('d'),
	('julianyear'))
SELECT
  l1.u, l2.u, l1.u::unit @ l2.u
FROM
  l l1 CROSS JOIN l l2
\crosstabview
     u      |     s      |          min           |           h            |           d            |           julianyear            
------------+------------+------------------------+------------------------+------------------------+---------------------------------
 s          | 1 s        | 0.0166666666666667 min | 0.000277777777777778 h | 1.15740740740741e-05 d | 3.16880878140289e-08 julianyear
 min        | 60 s       | 1 min                  | 0.0166666666666667 h   | 0.000694444444444444 d | 1.90128526884174e-06 julianyear
 h          | 3600 s     | 60 min                 | 1 h                    | 0.0416666666666667 d   | 0.000114077116130504 julianyear
 d          | 86400 s    | 1440 min               | 24 h                   | 1 d                    | 0.0027378507871321 julianyear
 julianyear | 31557600 s | 525960 min             | 8766 h                 | 365.25 d               | 1 julianyear
(5 rows)

